runtime/compiler/env/CHTable.cpp

```diff
@@ -991,6 +998,115 @@ bool CollectCompiledImplementors::visitSubclass(TR_PersistentClassInfo *cl)
    return false;
    }
 
+#if defined(J9VM_OPT_JITSERVER)
+class CollectResolvedImplementors: public CollectImplementors
+   {
+public:
+   CollectResolvedImplementors(TR::Compilation * comp,
+                           TR_OpaqueClassBlock *topClassId,
+                           int32_t maxCount,
+                           TR_ResolvedMethod *callerMethod,
+                           int32_t slotOrIndex,
+                           TR_YesNoMaybe useGetResolvedInterfaceMethod = TR_maybe) :
+      CollectImplementors(comp, topClassId, NULL, maxCount, callerMethod, slotOrIndex, useGetResolvedInterfaceMethod)
+      {
+      _implArray = new (comp->trHeapMemory()) TR_ResolvedMethod *[maxCount];
+      }
+
+   virtual bool visitSubclass(TR_PersistentClassInfo *cl) override;
+   void cacheResolvedMethods();
+
+private:
+   std::vector<TR_OpaqueClassBlock *> _subClasses;
+   };
+
+bool
+CollectResolvedImplementors::visitSubclass(TR_PersistentClassInfo *cl)
+   {
+   // must only be used on JITServer server-side
+   if (!comp()->isOutOfProcessCompilation())
+      {
+      stopTheWalk();
+      return false;
+      }
+
+   auto callerMethod = static_cast<TR_ResolvedJ9JITServerMethod *>(_callerMethod);
+   TR_ResolvedMethodType type = isInterface() ?
+      TR_ResolvedMethodType::Interface :
+      TR_ResolvedMethodType::VirtualFromOffset;
+   auto compInfoPT = static_cast<TR::CompilationInfoPerThreadRemote *>(static_cast<TR_J9VMBase *>(fe())->_compInfoPT);
+
+   // verify that our subclass meets all conditions
+   TR_OpaqueClassBlock *classId = cl->getClassId();
+   if (!TR::Compiler->cls.isAbstractClass(comp(), classId)
+      && !TR::Compiler->cls.isInterfaceClass(comp(), classId))
+      {
+      ++_numVisitedSubClasses;
+      if (_numVisitedSubClasses > _maxNumVisitedSubClasses)
+         {
+         // set count greater than maxCount, to indicate failure and force exit
+         _count = _maxCount + 1;
+         stopTheWalk();
+         return false;
+         }
+
+      // if the implementor hasn't been cached yet,
+      // add the corresponding subclass to the list
+      TR_ResolvedMethod *resolvedMethod;
+      TR_ResolvedMethodKey key =
+         compInfoPT->getResolvedMethodKey(
+            type,
+            reinterpret_cast<TR_OpaqueClassBlock *>(callerMethod->constantPoolHdr()),
+            _slotOrIndex,
+            classId);
+      if (!compInfoPT->getCachedResolvedMethod(
+             key,
+             callerMethod,
+             &resolvedMethod))
+         {
+         _subClasses.push_back(classId);
+         }
+      else if (resolvedMethod)
```

Comment:
Is it possible to find something in the cache, but `resolvedMethod` is NULL? If so, what should we do?

Reply:
Actually yes, in the case that we have found a cached unresolved method, we should immediately stop the walk and not make further remote calls. Added it.